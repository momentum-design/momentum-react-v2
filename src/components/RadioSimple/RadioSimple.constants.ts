const CLASS_PREFIX = 'md-radio-simple';

const DEFAULTS = {
  DISABLED: false,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
