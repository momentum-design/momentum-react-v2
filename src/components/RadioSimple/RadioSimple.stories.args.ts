import { commonStyles } from '../../storybook/helper.stories.argtypes';

const radioSimpleArgTypes = {
  children: {
    description:
      'Children is a prop that is used to render the RadioSimple components inside the RadioSimpleGroup.',
    control: { type: 'ReactNode', required: false },
    table: {
      type: {
        summary: 'ReactNode',
      },
      defaultValue: undefined,
    },
  },
  value: {
    description:
      'The value for the RadioSimple. This is used to provide a value for the RadioSimple.',
    control: { type: 'text', required: true },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: undefined,
    },
  },
  ariaLabel: {
    description:
      'The ariaLabel for the RadioSimple. This is used to provide an aria-label for the RadioSimple.',
    control: { type: 'text', required: false },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: 'example-ariaLabel',
    },
  },
  ariaLabelledBy: {
    description:
      'The ariaLabelledBy for the RadioSimple. This is used to provide an aria-labelledby for the RadioSimple.',
    control: { type: 'text', required: false },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: 'example-ariaLabelledBy',
    },
  },
  isDisabled: {
    description:
      'The isDisabled for the RadioSimple. This is typically set the disable state for the RadioSimple.',
    control: { type: 'boolean', required: false },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: false,
    },
  },
};

export { radioSimpleArgTypes };

export default {
  ...commonStyles,
  ...radioSimpleArgTypes,
};
