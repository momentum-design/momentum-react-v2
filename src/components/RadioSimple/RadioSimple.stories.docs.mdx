The `<RadioSimple />` component.

Extended `AriaRadioProps`. See
[React aria useRadioGroup](https://react-spectrum.adobe.com/react-aria/useRadioGroup.html)
