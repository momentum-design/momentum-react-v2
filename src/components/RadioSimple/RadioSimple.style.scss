.md-radio-simple-wrapper {
  cursor: pointer;
  display: flex;
  border-radius: 0.25rem;
  width: 6rem;
  background-color: var(--mds-color-theme-background-primary-ghost);

  &:hover {
    background-color: var(--mds-color-theme-background-primary-hover);
  }

  &:active {
    background-color: var(--mds-color-theme-background-primary-active);
  }

  &[data-selected='true'] {
    background-color: var(--mds-color-theme-background-primary-active);
  }

  &[data-focused='true'] {
    border-color: transparent;
    outline: none;
    box-shadow: var(--md-globals-focus-ring-box-shadow);
  }

  &[data-disabled='true'] {
    cursor: default;
    background-color: var(--mds-color-theme-background-primary-disabled);
  }
}
