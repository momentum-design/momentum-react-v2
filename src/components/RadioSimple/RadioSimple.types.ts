import { CSSProperties, ReactNode } from 'react';
import { AriaRadioProps } from '@react-types/radio';

export interface RadioSimpleProps extends AriaRadioProps {
  /**
   * Optional ariaLabel on the RadioSimple component.
   */
  ariaLabel?: string;

  /**
   * Optional ariaLabelledBy on the RadioSimple component.
   */
  ariaLabelledBy?: string;

  /**
   * Child components of this RadioSimple.
   */
  children?: ReactNode;

  /**
   * Custom class for overriding this component's CSS.
   */
  className?: string;

  /**
   * The id for the RadioSimple component.
   */
  id?: string;

  /**
   * Disabled state for the RadioSimple component.
   */
  isDisabled?: boolean;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;

  /**
   * The value of the RadioSimple component.
   */
  value: string;
}
