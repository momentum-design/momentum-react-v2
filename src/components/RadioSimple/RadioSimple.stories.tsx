import React from 'react';
import { Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';
import argTypes from './RadioSimple.stories.args';

import RadioSimple, { RadioSimpleProps } from '.';

import Documentation from './RadioSimple.stories.docs.mdx';
import RadioSimpleGroup from '../RadioSimpleGroup';
import Text from '../Text';
import Icon from '../Icon';

export default {
  title: 'Momentum UI/RadioSimple',
  component: RadioSimple,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

const Example = Template<RadioSimpleProps>((args: RadioSimpleProps) => {
  return (
    <RadioSimpleGroup label="Example" description="Value of the group is 'red'" value={'red'}>
      <RadioSimple {...args} />
    </RadioSimpleGroup>
  );
}).bind({});

Example.argTypes = { ...argTypes };

Example.args = {
  children: (
    [
      <Icon key="1" name="accessibility" autoScale />,
      <Text key="2">My value is 'red'</Text>
    ]
  ),
  className: 'example-className',
  id: 'example-id',
  style: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    border: '1px solid red',
    padding: '8px',
    margin: '8px 8px'
  },
};

export { Example };
