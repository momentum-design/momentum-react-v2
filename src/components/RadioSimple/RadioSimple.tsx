import React, { FC, useContext, useRef } from 'react';
import classnames from 'classnames';
import { useRadio } from '@react-aria/radio';
import { VisuallyHidden } from '@react-aria/visually-hidden';
import { useFocusRing } from '@react-aria/focus';
import { mergeProps } from '@react-aria/utils';

import { STYLE, DEFAULTS } from './RadioSimple.constants';
import { RadioSimpleProps } from './RadioSimple.types';
import './RadioSimple.style.scss';
import { RadioSimpleGroupContext } from '../RadioSimpleGroup';

/**
 * The RadioSimple component.
 */
const RadioSimple: FC<RadioSimpleProps> = (props: RadioSimpleProps) => {
  const {
    ariaLabel,
    ariaLabelledBy,
    children,
    className,
    id,
    isDisabled = DEFAULTS.DISABLED,
    style,
    value,
  } = props;
  const state  = useContext(RadioSimpleGroupContext);
  const ref = useRef(null);
  const isSelected = state?.selectedValue === value;
  const { focusProps, isFocusVisible } = useFocusRing();
  const { inputProps } = useRadio(
    { 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy, ...props },
    state,
    ref
  );

  return (
    <label
      className={classnames(STYLE.wrapper, className)}
      data-focused={isFocusVisible}
      data-selected={isSelected}
      data-disabled={isDisabled}
      style={style}
      id={id}
    >
      <VisuallyHidden>
        <input {...mergeProps(inputProps, focusProps)} ref={ref} />
      </VisuallyHidden>
      {children}
    </label>
  );
};

export default RadioSimple;
