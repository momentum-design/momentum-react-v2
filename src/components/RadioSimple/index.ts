import RadioSimple from './RadioSimple';

export * as RADIO_SIMPLE_CONSTANTS from './RadioSimple.constants';
export * from './RadioSimple.types';
export default RadioSimple;
