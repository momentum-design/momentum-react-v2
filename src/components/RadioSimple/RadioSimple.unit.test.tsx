import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import RadioSimple from '.';
import { STYLE } from './RadioSimple.constants';
import { RadioSimpleGroupContext } from '../RadioSimpleGroup';
import Text from '../Text';
import Icon from '../Icon';

describe('<RadioSimple />', () => {
  const children = (
    <div>
      <Icon name="accessibility" autoScale />
      <Text>Red</Text>
    </div>
  );

  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const { asFragment } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red'/>
        </RadioSimpleGroupContext.Provider>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should match snapshot with ariaLabel', () => {
      expect.assertions(1);

      const ariaLabel = 'example-ariaLabel';

      const { asFragment } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' ariaLabel={ariaLabel}/>
        </RadioSimpleGroupContext.Provider>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should match snapshot with ariaLabelledBy', () => {
      expect.assertions(1);

      const ariaLabelledBy = 'example-ariaLabel';

      const { asFragment } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' ariaLabelledBy={ariaLabelledBy}/>
        </RadioSimpleGroupContext.Provider>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-className';

      const { asFragment } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' className={className}/>
        </RadioSimpleGroupContext.Provider>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const { asFragment } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' id={id}/>
        </RadioSimpleGroupContext.Provider>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should match snapshot with isDisabled', () => {
      expect.assertions(1);

      const isDisabled = true;

      const { asFragment } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red'isDisabled={isDisabled}/>
        </RadioSimpleGroupContext.Provider>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'red' };

      const { asFragment } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' style={style}/>
        </RadioSimpleGroupContext.Provider>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its default wrapper class', () => {
      expect.assertions(1);

      const { container } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red'/>
        </RadioSimpleGroupContext.Provider>
      );

      const radioSimple = container.querySelector('label');

      expect(radioSimple).toHaveClass(STYLE.wrapper);
    });

    it('should have provided ariaLabel when ariaLabel is provided', () => {
      expect.assertions(1);

      const ariaLabel = 'example-ariaLabel';

      const { container } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' ariaLabel={ariaLabel}/>
        </RadioSimpleGroupContext.Provider>
      );

      const radioSimple = container.querySelector('input');

      expect(radioSimple.getAttribute('aria-label')).toBe(ariaLabel);
    });

    it('should have provided ariaLabel when ariaLabelledBy is provided', () => {
      expect.assertions(1);

      const ariaLabelledBy = 'example-ariaLabelledBy';

      const { container } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' ariaLabelledBy={ariaLabelledBy}/>
        </RadioSimpleGroupContext.Provider>
      );

      const radioSimple = container.querySelector('input');

      expect(radioSimple.getAttribute('aria-labelledby')).toBe(ariaLabelledBy);
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-className';

      const { container } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' className={className}/>
        </RadioSimpleGroupContext.Provider>
      );

      const radioSimple = container.querySelector('label');

      expect(radioSimple).toHaveClass(className);
    });

    it('should have provided id when id is provided', () => {
      expect.assertions(1);

      const id = 'example-id';

      const { container } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' id={id}/>
        </RadioSimpleGroupContext.Provider>
      );

      const radioSimple = container.querySelector('label');

      expect(radioSimple.getAttribute('id')).toBe(id);
    });

    it('should have provided isDisabled when isDisabled is provided', () => {
      expect.assertions(1);

      const isDisabled = true;

      const { container } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' isDisabled={isDisabled}/>
        </RadioSimpleGroupContext.Provider>
      );

      const radioSimple = container.querySelector('label');

      expect(radioSimple.getAttribute('data-disabled')).toBe('true');
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const { container } = render(
        <RadioSimpleGroupContext.Provider value={'red'}>
          <RadioSimple children={children} value='red' style={style}/>
        </RadioSimpleGroupContext.Provider>
      );

      const radioSimple = container.querySelector('label');

      expect(radioSimple).toHaveStyle(style);
    });
  });
});
