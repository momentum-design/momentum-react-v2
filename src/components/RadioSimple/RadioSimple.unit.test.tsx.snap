// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioSimple /> snapshot should match snapshot 1`] = `
<DocumentFragment>
  <label
    class="md-radio-simple-wrapper"
    data-disabled="false"
    data-focused="false"
    data-selected="false"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        tabindex="0"
        type="radio"
        value="red"
      />
    </div>
    <div>
      <p
        class="md-text-wrapper"
        data-type="body-primary"
      >
        Red
      </p>
    </div>
  </label>
</DocumentFragment>
`;

exports[`<RadioSimple /> snapshot should match snapshot with ariaLabel 1`] = `
<DocumentFragment>
  <label
    class="md-radio-simple-wrapper"
    data-disabled="false"
    data-focused="false"
    data-selected="false"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        aria-label="example-ariaLabel"
        tabindex="0"
        type="radio"
        value="red"
      />
    </div>
    <div>
      <p
        class="md-text-wrapper"
        data-type="body-primary"
      >
        Red
      </p>
    </div>
  </label>
</DocumentFragment>
`;

exports[`<RadioSimple /> snapshot should match snapshot with ariaLabelledBy 1`] = `
<DocumentFragment>
  <label
    class="md-radio-simple-wrapper"
    data-disabled="false"
    data-focused="false"
    data-selected="false"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        aria-labelledby="example-ariaLabel"
        tabindex="0"
        type="radio"
        value="red"
      />
    </div>
    <div>
      <p
        class="md-text-wrapper"
        data-type="body-primary"
      >
        Red
      </p>
    </div>
  </label>
</DocumentFragment>
`;

exports[`<RadioSimple /> snapshot should match snapshot with className 1`] = `
<DocumentFragment>
  <label
    class="md-radio-simple-wrapper example-className"
    data-disabled="false"
    data-focused="false"
    data-selected="false"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        tabindex="0"
        type="radio"
        value="red"
      />
    </div>
    <div>
      <p
        class="md-text-wrapper"
        data-type="body-primary"
      >
        Red
      </p>
    </div>
  </label>
</DocumentFragment>
`;

exports[`<RadioSimple /> snapshot should match snapshot with id 1`] = `
<DocumentFragment>
  <label
    class="md-radio-simple-wrapper"
    data-disabled="false"
    data-focused="false"
    data-selected="false"
    id="example-id"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        id="example-id"
        tabindex="0"
        type="radio"
        value="red"
      />
    </div>
    <div>
      <p
        class="md-text-wrapper"
        data-type="body-primary"
      >
        Red
      </p>
    </div>
  </label>
</DocumentFragment>
`;

exports[`<RadioSimple /> snapshot should match snapshot with isDisabled 1`] = `
<DocumentFragment>
  <label
    class="md-radio-simple-wrapper"
    data-disabled="true"
    data-focused="false"
    data-selected="false"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        disabled=""
        type="radio"
        value="red"
      />
    </div>
    <div>
      <p
        class="md-text-wrapper"
        data-type="body-primary"
      >
        Red
      </p>
    </div>
  </label>
</DocumentFragment>
`;

exports[`<RadioSimple /> snapshot should match snapshot with style 1`] = `
<DocumentFragment>
  <label
    class="md-radio-simple-wrapper"
    data-disabled="false"
    data-focused="false"
    data-selected="false"
    style="color: red;"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        tabindex="0"
        type="radio"
        value="red"
      />
    </div>
    <div>
      <p
        class="md-text-wrapper"
        data-type="body-primary"
      >
        Red
      </p>
    </div>
  </label>
</DocumentFragment>
`;
